File: Problems/0015.3sum/threeSum.go
package leetcode

import "sort"

// Runtime: 84 ms, faster than 35.65% of Go online submissions
func threeSum(nums []int) [][]int {
	counter := make(map[int]int)
	res := make([][]int, 0)

	for _, n := range nums {
		counter[n]++
	}
	uniqNums := []int{}
	for k := range counter {
		uniqNums = append(uniqNums, k)
	}
	sort.Ints(uniqNums)

	for i := range uniqNums {
		for j := i + 1; j < len(uniqNums); j++ {
			if uniqNums[i]*2+uniqNums[j] == 0 && counter[uniqNums[i]] >= 2 {
				res = append(res, []int{uniqNums[i], uniqNums[i], uniqNums[j]})
			} else if uniqNums[i]+uniqNums[j]*2 == 0 && counter[uniqNums[j]] >= 2 {
				res = append(res, []int{uniqNums[i], uniqNums[j], uniqNums[j]})
			} else {
				r := 0 - uniqNums[i] - uniqNums[j]
				if counter[r] > 0 && r > uniqNums[j] {
					res = append(res, []int{uniqNums[i], uniqNums[j], r})
				}
			}

		}
	}
	if counter[0] >= 3 {
		res = append(res, []int{0, 0, 0})
	}
	return res
}

// Runtime: 28 ms, faster than 94.38% of Go online submissions
func threeSumV2(nums []int) [][]int {
	res := [][]int{}
	sort.Ints(nums)
	length := len(nums)

	for i := 0; i < length; i++ {
		for j, k := i+1, length-1; j < k; {
			sum := nums[i] + nums[j] + nums[k]
			if sum > 0 {
				k--
			} else if sum == 0 {
				res = append(res, []int{nums[i], nums[j], nums[k]})
				for j+1 < length && j < k && nums[j] == nums[j+1] {
					j++
				}
				j++
			} else {
				j++
			}
		}
		for i+1 < length && nums[i] == nums[i+1] {
			i++
		}
	}
	return res
}

File: Problems/0015.3sum/threeSum.py
from typing import List, Tuple

class Solution:

    # Runtime: 360 ms, faster than 98.51% of Python3 online submissions for 3Sum.
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        pos_counter = {}
        neg_counter = {}
        for n in nums:
            if n >= 0:
                pos_counter[n] = pos_counter.get(n, 0) + 1
            else:
                neg_counter[n] = neg_counter.get(n, 0) + 1


        for pos in pos_counter:
            for neg in neg_counter:
                x = 0 - pos - neg
                if x in pos_counter or x in neg_counter:

                    if (x == pos and pos_counter[x] > 1) or \
                        (x == neg and neg_counter[x] > 1):
                        res.append([neg, x, pos])
                    elif x > pos:
                        res.append([neg, pos, x])
                    elif x < neg:
                        res.append([x, neg, pos])

        if pos_counter.get(0, 0) >= 3:
            res.append([0, 0, 0])
        return res

    def threeSumV2(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        counter = {}
        for n in nums:
            counter[n] = counter.get(n, 0) + 1

        length = len(nums)
        for i in range(length - 2):
            fst = nums[i]
            sec = nums[i + 1]
            
            for j in range(i+2, length-1):
                nums[i]